server/api-spec-1.0.yaml

```diff
@@ -1,298 +1,298 @@
-      type: object
-      properties:
-        sender:
-          $ref: '#/components/schemas/Username'
-        receiver:
-          $ref: '#/components/schemas/Username'
-        amount:
-          type: integer
-          description: The amount to transfer, in cents
-          example: 1337
-
-    Deposit:
-      type: object
-      properties:
-        user:
-          $ref: '#/components/schemas/Username'
-        amount:
-          type: integer
-          minimum: 1
-          description: Amount to be deposited
-        type:
-          type: string
-          description: The type of deposit to be made
-          # TODO: make a new API to fetch the list of supported payments
-          enum:
-            - cash
-            - card
+openapi: '3.0.2'
+info:
+  title: PixelBank API
+  version: '0.0.1'
+servers:
+  - url: https://localhost:8080
+# TODOs:
+# - add auth for admin & protected users?
+#   - if so, add new 401 errors for APIs
+# - add API for fetching supported payment types
+# - work out a card payment flow then abstract that to all payment flows (i.e. ask bank to make a request, get a response with some action, confirm it)
+# - document all the things!
+tags:
+  - name: User
+    description: Interact with Users of the Bank
+  - name: Admin
+    description: Administrate the bank
+  - name: Transactions
+    description: Create transactions
+  - name: Utilities
+    description: Utilities for checking liveness, uptime, etc
+paths:
+  /ping:
+    get:
+      summary: Check if PixelBank is responding
+      description: Returns status code 200, provided PixelBank is up and responding to requests
+      tags:
+        - Utilities
+      responses:
+        '200':
+          description: OK
+  ### USERS
+  /users/:
+    post:
+      summary: Create new user
+      description: Create a new user
+      tags:
+        - User
+      requestBody:
+        description: |
+          Details of the user to create.
+
+          Note:
+
+          * Users cannot be created with a negative balance
+          * Admin users cannot be created via the API
+        content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/CreateUser'
+        required: true
+      responses:
+        '200':
+          description: Success
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/User'
+        '400':
+          description: Invalid user object
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/User'
+# TODO: Create API Error response thing and use it here
+        '403':
+          description: Forbidden
+        '409':
+          description: Resource already exists
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/Error'
+
+
+  /users/{userID}:
+    parameters:
+        - name: username

```

Comment:
```suggestion
        - name: userID

```

Reply:
Ah yes, I need to fix these, but I should get your opinion on this too:
I split the User get into two calls, the regular one fetching a user by their id, and a users/name call to fetch them by username. What's your opinion on that?

Reply:
Personally, I think we should do all user operations on username as they're guaranteed unique anyway (but we should probably handle normalizing for case matches etc!). Simplifies it for us, and we don't have to support two versions of the API (because if we offer one API that handles IDs we should probably offer them all). 

FWIW, I experimented a little with [Utoipa](https://docs.rs/utoipa/latest/utoipa/index.html) to have the API spec live next to the code implementing it, and (if you agree) I'd like to start working on moving the API spec over to this format sooner rather than later (like starting tonight if I have time) so we don't have to spend forever doing it down the road.

Reply:
Working that way does force usernames to be unchangeable in the future, where we're not prone to a race condition when a user changes their username while a series of calls is being made. Sure, they're both unique but I'd prefer to keep the username changeable and the user ID not. It's just safer that way in my opinion.

Reply:
I do like the introduction of Utopia, but I really want to close this merge request first. Can I just leave the current openapi file as is and you just nuke it when you implement utoipa?

Reply:
> Working that way does force usernames to be unchangeable in the future, where we're not prone to a race condition when a user changes their username while a series of calls is being made. Sure, they're both unique but I'd prefer to keep the username changeable and the user ID not. It's just safer that way in my opinion.

This is a good point I hadn't considered. Let's chat a little bit tonight once I've had some time to think.

Reply:
> I do like the introduction of Utopia, but I really want to close this merge request first. Can I just leave the current openapi file as is and you just nuke it when you implement utoipa?

Yes totally feel free to close them out - let's get this merged today once the other issues are resolved :)

Reply:
> > Working that way does force usernames to be unchangeable in the future, where we're not prone to a race condition when a user changes their username while a series of calls is being made. Sure, they're both unique but I'd prefer to keep the username changeable and the user ID not. It's just safer that way in my opinion.
> 
> This is a good point I hadn't considered. Let's chat a little bit tonight once I've had some time to think. 

Sorry, I'm not going to be in the space tonight, my parents are dropping off the cat tonight :)

Reply:
> > I do like the introduction of Utopia, but I really want to close this merge request first. Can I just leave the current openapi file as is and you just nuke it when you implement utoipa?
> 
> Yes totally feel free to close them out - let's get this merged today once the other issues are resolved :) 

I want to say that they all are, so we should be able to merge now :D

Reply:
> > > Working that way does force usernames to be unchangeable in the future, where we're not prone to a race condition when a user changes their username while a series of calls is being made. Sure, they're both unique but I'd prefer to keep the username changeable and the user ID not. It's just safer that way in my opinion.
> > 
> > 
> > This is a good point I hadn't considered. Let's chat a little bit tonight once I've had some time to think.
> 
> Sorry, I'm not going to be in the space tonight, my parents are dropping off the cat tonight :)

No worries, we can chat here or on discord some other time :) 



> > > I do like the introduction of Utopia, but I really want to close this merge request first. Can I just leave the current openapi file as is and you just nuke it when you implement utoipa?
> > 
> > 
> > Yes totally feel free to close them out - let's get this merged today once the other issues are resolved :)
> 
> I want to say that they all are, so we should be able to merge now :D

Only one left is this one: https://github.com/NinjaLikesCheez/PixelBank/pull/7#issuecomment-1584700086 then we're good to go :)

Reply:
(Resolving this, we can chat about https://github.com/NinjaLikesCheez/PixelBank/pull/7#issuecomment-1584700086 over there if needed!)